src/types.c

```diff
@@ -42,12 +46,60 @@ const char* vec_type_as_str(enum vctrs_type type) {
   case vctrs_type_raw:       return "raw";
   case vctrs_type_list:      return "list";
   case vctrs_type_dataframe: return "dataframe";
+  case vctrs_type_s3:        return "s3";
+  case vctrs_type_scalar:    return "scalar";
+  }
+}
+
+static SEXP vec_is_vector_dispatch_fn = NULL;
+
+bool vec_is_vector(SEXP x) {
+  switch (vec_typeof(x)) {
+  case vctrs_type_null:
+  case vctrs_type_scalar:
+    return false;
+  case vctrs_type_logical:
+  case vctrs_type_integer:
+  case vctrs_type_double:
+  case vctrs_type_complex:
+  case vctrs_type_character:
+  case vctrs_type_raw:
+  case vctrs_type_list:
+  case vctrs_type_dataframe:
+    return true;
+  case vctrs_type_s3:
+    if (Rf_inherits(x, "vctrs_vctr")) {
+      return true;
+    } else {
+      SEXP dispatch_call = PROTECT(Rf_lang2(vec_is_vector_dispatch_fn, x));
+      SEXP out = Rf_eval(dispatch_call, R_GlobalEnv);
```

Comment:
Think about this more — we'll eventually need to write something about our "internal" generics, and how in return for performance we sacrifice a small amount of correctness, so that methods locally defined methods don't work. That tradeoff is worth making because the only place we routinely use local methods is in testing, and we have work arounds for that.